ignite/src/main/java/com/yahoo/ycsb/db/IgniteYCSBClient.java

```diff
@@ -0,0 +1,184 @@
+    Properties props = getProperties();
+    tableName = props.getProperty(CoreWorkload.TABLENAME_PROPERTY, CoreWorkload.TABLENAME_PROPERTY_DEFAULT);
+    String igniteIPs = props.getProperty(IGNITE_IPS, "127.0.0.1,127.0.0.1:47500..47509");
+    asyncInsert = Boolean.valueOf(props.getProperty(INSERT_ASYNC, "false"));
+    batchSize = Integer.parseInt(props.getProperty(BATCH_SIZE, "1"));
+    createIgniteClient(tableName, igniteIPs);
+
+    LOG.info(
+        "" + this +  ":" +
+        "\n\tBATCH_SIZE = " + batchSize +
+        "\n\tINSERT_ASYNC = " +asyncInsert);
+
+  }
+
+  @Override
+  public Status read(String table, String key, Set<String> fields,
+                     Map<String, ByteIterator> result) {
+    LOG.debug("IgniteYCSBClient.read:{}", key);
+    Map<String, String> r = cache.get(key);
+    if (r == null) {
+      return Status.NOT_FOUND;
+    } else {
+      populateResult(result, fields, r);
+      return Status.OK;
+    }
+  }
+
+  private void populateResult(Map<String, ByteIterator> sink, Set<String> fields, Map<String, String> source) {
+    if (fields == null) {
+      for (Map.Entry<String, String> entry : source.entrySet()) {
+        sink.put(entry.getKey(), new StringByteIterator(entry.getValue()));
+      }
+    } else {
+      for (String fName : fields) {
+        sink.put(fName, new StringByteIterator(source.get(fName)));
+      }
+    }
+  }
+
+  @Override
+  public Status scan(String table, String startkey, int recordcount,
+                     Set<String> fields,
+                     Vector<HashMap<String, ByteIterator>> result) {
+    LOG.warn("Ignite does not support scan semantics");
+    return Status.NOT_IMPLEMENTED;
+  }
+
+  @Override
+  public Status update(String table, String key,
+                       Map<String, ByteIterator> values) {
+    try {
+      cache.put(key, toCachedValue(values));
+      return Status.OK;
+    } catch (Exception e) {
+      LOG.error("update error", e);
+      return Status.ERROR;
+    }
+  }
+
+
+  /**
+   * @param table  The name of the table
+   * @param key    The record key of the record to insert.
+   * @param values A HashMap of field/value pairs to insert in the record
+   * @return
+   */
+
+  @Override
+  public Status insert(String table, String key,
+                       Map<String, ByteIterator> values) {
+    LOG.debug("IgniteYCSBClient.insert:{}|{}", table, values);
+    Map<String, String> valToCache = toCachedValue(values);
+    try {
+
+      if (batchSize > 1) {
+        batchBuffer.put(key, valToCache);
+        if (++batchCount % batchSize == 0) {
+          insertBuffer();
+        }
+
+      } else {
+        if (asyncInsert) {
+          cache.putAsync(key, valToCache);
+        } else {
+          cache.put(key, valToCache);
+        }
+      }
+
+
+      return Status.OK;
+    } catch (Exception e) {
+      LOG.error("insert error", e);
+      return Status.ERROR;
+    }
+  }
+
+  @Override
+  public void cleanup() throws DBException {
```

Comment:
We need some way to ensure that once this method returns any outstanding Async calls have completed. Looking at the [IgniteCache API](https://ignite.apache.org/releases/latest/javadoc/org/apache/ignite/IgniteCache.html), I don't see an obvious way to do this besides collecting the relevant IgniteFuture objects as we go.